src/db/couchUsers.ts

```diff
@@ -0,0 +1,183 @@
+import { asBoolean, asMap, asObject, asOptional, Cleaner } from 'cleaners'
+import { asCouchDoc, asMaybeNotFoundError } from 'edge-server-tools'
+import { ServerScope } from 'nano'
+
+import {
+  Device,
+  User,
+  UserCurrencyCodes,
+  UserCurrencyHours,
+  UserDevices,
+  UserNotifications
+} from '../types/pushTypes'
+import { asCouchDevice, unpackDevice } from './couchDevices'
+import { devicesSetup, usersSetup } from './couchSetup'
+import { saveToDB } from './utils/couchOps'
+
+export const asUserDevices: Cleaner<UserDevices> = asObject(asBoolean)
+export type IDevicesByCurrencyHoursViewResponse = ReturnType<
+  typeof asUserDevices
+>
+export const asUserCurrencyHours: Cleaner<UserCurrencyHours> = asObject({
+  '1': asBoolean,
+  '24': asBoolean
+})
+export const asUserCurrencyCodes: Cleaner<UserCurrencyCodes> =
+  asMap(asUserCurrencyHours)
+
+export const asUserNotifications: Cleaner<UserNotifications> = asObject({
+  enabled: asOptional(asBoolean),
+  currencyCodes: asUserCurrencyCodes
+})
+
+export const asCouchUser = asCouchDoc<Omit<User, 'userId'>>(
+  asObject({
+    devices: asUserDevices,
+    notifications: asUserNotifications
+  })
+)
+
+type CouchUser = ReturnType<typeof asCouchUser>
+
+// ------------------------------------------------------------------------------
+// Functions associated with User
+// ------------------------------------------------------------------------------
+
+export const registerNotifications = async (
+  userId: string,
+  currencyCodes: string[],
+  connection: ServerScope
+): Promise<void> => {
+  const db = connection.db.use(usersSetup.name)
+  const raw = await db.get(userId).catch(error => {
+    if (asMaybeNotFoundError(error) != null) return
+    throw error
+  })
+  const user = asCouchUser(raw)
+
+  const currencyCodesToUnregister = Object.keys(
+    user.doc.notifications.currencyCodes
+  ).filter(code => !currencyCodes.includes(code))
+  for (const code of currencyCodesToUnregister) {
+    delete user.doc.notifications.currencyCodes[code]
+  }
+
+  for (const code of currencyCodes) {
+    if (code in user.doc.notifications.currencyCodes) continue
+
+    user.doc.notifications.currencyCodes[code] = {
+      '1': true,
+      '24': true
+    }
+  }
+  await saveToDB(db, user, asCouchUser)
+}
+export const devicesByCurrencyHours = async (
+  hours: string,
+  currencyCode: string,
+  connection: ServerScope
+) => {
+  return await connection.db
+    .use(usersSetup.name)
+    .view<IDevicesByCurrencyHoursViewResponse>('filter', 'by-currency', {
+      key: [currencyCode, hours]
+    })
+}
```

Comment:
It would be interesting to see how many rows this returns for a query like "24". I'm guessing it can be in the millions. We might want to use the streaming view stuff from edge-server-tools to turn this into a generator. https://github.com/EdgeApp/edge-server-tools/blob/master/docs/view-to-stream.md